### PR TITLE
Try a new way of splitting parmap input into chunks.

### PR DESCRIPTION
**Pull request by mjambon**
I tried to explain the reasoning in the comments. Let's see how this affects the benchmarks.

PR checklist:
- [x] documentation is up to date
- [x] changelog is up to date

### CODE REVIEW

**Comment by emjin** [comment]
I'll update our benchmark version after next release

**Comment by emjin** [comment]
Re: speed testing - I saw these speedups in another PR, so unclear what affect this PR had. We will need to test it locally, and improve our benchmark suite based on that work. To be fair, I think these benchmarks used to be better---we've just gotten faster on them.

**Comment by mjambon** [comment]
I updated the implementation, which now simply specifies `chunksize:1`. This is the best granularity we can get without pre-sorting the jobs.

**Comment by mjambon** [comment]
The next thing we could do is sort the input targets by decreasing size, hoping that the size correlates with run time. This would allow parmap to finish with the smaller jobs and somewhat maximize CPU usage. I believe it's a type of greedy algorithm and I guess it's not optimal but close enough.

**Comment by mjambon** [comment]
I just added the sort-by-decreasing size.

**Review by IagoAbal — APPROVED** [review]
Nice! Did you see some speed up? I tried this myself although I did in a more hacky way just to experiment, and I didn't see much improvement on one benchmark that I've been using recently. It could be because I was using semgrep-core directly and it was filtering out the minimized JS files.

File: semgrep-core/src/core/Find_target.ml [hunk 1]
```diff
@@ -142,7 +142,14 @@ let filter_by_size _lang paths =
              }
          else Ok path)
 
-let files_of_dirs_or_files ?(keep_root_files = true) lang roots =
+let sort_by_decreasing_size paths =
+  paths
+  |> List.map (fun path -> (path, Common2.filesize path))
```
Comment (IagoAbal) [inline semgrep-core/src/core/Find_target.ml]: `Common.map` ?

Comment (IagoAbal) [reply semgrep-core/src/core/Find_target.ml]: Actually, here `List.rev_map` works fine.

File: semgrep-core/src/core/Find_target.ml [hunk 2]
```diff
@@ -142,7 +142,14 @@ let filter_by_size _lang paths =
...
+  |> List.sort (fun (_, (a : int)) (_, b) -> compare b a)
+  |> List.map fst
```
Comment (IagoAbal) [inline semgrep-core/src/core/Find_target.ml]: `Common.map`

**Comment by IagoAbal** [comment]
I'm guessing the tests fail because sorting the files affected the output. You need to `make regenerate-tests`.

**Comment by mjambon** [comment]
I'll resume next week. We should run the full benchmarks before merging.

**Comment by mjambon** [comment]
I fixed a failing test by sorting the results correctly but I'm still getting an error locally:
```
~/semgrep/semgrep $ pytest 'tests/qa/test_public_repos.py::test_semgrep_on_repo[repo_object0]' -vv --snapshot-update
============================= test session starts ==============================
platform linux -- Python 3.8.10, pytest-6.2.5, py-1.10.0, pluggy-1.0.0 -- /home/martin/.local/share/virtualenvs/semgrep-JT-h3F_l/bin/python
cachedir: .pytest_cache
rootdir: /home/martin/semgrep/semgrep
plugins: xdist-2.3.0, snapshot-0.6.1, forked-1.3.0
collected 1 item                                                               

tests/qa/test_public_repos.py::test_semgrep_on_repo[repo_object0] FAILED [100%]

=================================== FAILURES ===================================
______________________ test_semgrep_on_repo[repo_object0] ______________________
[XPASS(strict)] https://github.com/tree-sitter/tree-sitter-ruby/issues/167
----------------------------- Captured stdout call -----------------------------
semgrep command: ['/home/martin/.local/share/virtualenvs/semgrep-JT-h3F_l/bin/python', '-m', 'semgrep', '--config=rules/regex-sentinel.yaml', '--strict', '--json', '--optimizations', 'none', PosixPath('/home/martin/.cache/semgrep-tests/rails-rails-latest')]
--- semgrep error output ---
running 1 rules...
ran 1 rules on 2915 files: 1 findings

----------------------------------------------------------------------------------------------
Semgrep sped up by 5x since version 0.51. We release weekly; update often!

Version 0.68 of Semgrep will change behavior of metrics when using Semgrep.dev Registry rules.
See https://github.com/returntocorp/semgrep/pull/3874 for more information.
----------------------------------------------------------------------------------------------

----------------------------
--- semgrep standard output ---
{"errors": [], "results": [{"check_id": "rules.test-regex", "end": {"col": 29, "line": 1, "offset": 28}, "extra": {"is_ignored": false, "lines": "sentinel = 87518275812375164", "message": "test regex message", "metadata": {}, "metavars": {}, "severity": "ERROR"}, "path": "/home/martin/.cache/semgrep-tests/rails-rails-latest/sentinel.rb", "start": {"col": 8, "line": 1, "offset": 7}}]}

-------------------------------
=========================== short test summary info ============================
FAILED tests/qa/test_public_repos.py::test_semgrep_on_repo[repo_object0]
============================== 1 failed in 26.78s ==============================
```

I have no idea what causes this test to fail for me. The function `test_semgrep_on_repo` in `test_public_repos.py` looks like it returns normally (printing something at the end of the function body is successful). Maybe @brendongo knows something about this.

**Comment by brendongo** [comment]
Looked into this. Looks like local running is using a stale version of the test target repo. `rm -rf /home/martin/.cache/semgrep-tests/rails-rails-latest` should cause your run to succeed locally

File: semgrep/setup.py [hunk 3]
```diff
@@ -92,6 +103,15 @@ def find_executable(env_name, exec_name):
...
+        # The semgrep-core executable doesn't have the write
+        # permission (because of something dune does?), and copyfile
+        # doesn't remove the destination file if it already exists
+        # but tries to truncate it, resulting in an error.
+        # So we remove the destination file first if it exists.
+        try:
+            os.remove(dst)
+        except OSError:
+            pass
         shutil.copyfile(src, dst)
```
Comment (mjambon) [inline semgrep/setup.py:115]: fixes error when running `make` from the project root

File: semgrep/semgrep/core_output.py [hunk 4]
```diff
@@ -353,7 +353,11 @@ def convert_to_rule_match(match: CoreMatch, rule: Rule) -> RuleMatch:
...
-        # Before this PR was sorting, keeping but not sure why
-        findings = sorted(findings, key=lambda rule_match: rule_match.start.offset)
+        # Sort results so as to guarantee the same results across different
+        # runs. Results may arrive in a different order due to parallelism
+        # (-j option).
+        findings = sorted(
+            findings, key=lambda rule_match: [rule_match.path, rule_match.start.offset]
+        )
         findings = dedup(findings)
         return findings
```
Comment (mjambon) [inline semgrep/semgrep/core_output.py:363]: fixes inconsistencies in test results due to semgrep running multiple jobs in parallel.

**Pull request merged**